src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -0,0 +1,571 @@
+-- |
+-- Module      : Streamly.Internal.Data.Stream.IsStream.Top
+-- Copyright   : (c) 2020 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Top level IsStream module that can use all other lower level IsStream
+-- modules.
+
+module Streamly.Internal.Data.Stream.IsStream.Top
+    (
+    -- * Transformation
+    -- ** Sampling
+    -- | Value agnostic filtering.
+      sampleFromThen
+    , sampleIntervalStart
+    , sampleIntervalEnd
+    , sampleBurstStart
+    , sampleBurstEnd
+
+    -- * Nesting
+    -- ** Set like operations
+    -- | These are not exactly set operations because streams are not
+    -- necessarily sets, they may have duplicated elements.
+    , intersectBy
+    , mergeIntersectBy
+    , differenceBy
+    , mergeDifferenceBy
+    , unionBy
+    , mergeUnionBy
+
+    -- ** Join operations
+    , crossJoin
+    , innerJoin
+    , mergeInnerJoin
+    , hashInnerJoin
+    , leftJoin
+    , mergeLeftJoin
+    , hashLeftJoin
+    , outerJoin
+    , mergeOuterJoin
+    , hashOuterJoin
+    )
+where
+
+#include "inline.hs"
+
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.State.Strict (get, put)
+-- import Data.Hashable (Hashable)
+import Data.IORef (newIORef, readIORef, modifyIORef')
+import Data.Kind (Type)
+#if !(MIN_VERSION_base(4,11,0))
+import Data.Semigroup (Semigroup(..))
+#endif
+import Streamly.Internal.Data.SVar (MonadAsync)
+import Streamly.Internal.Data.Stream.IsStream.Common (concatM)
+import Streamly.Internal.Data.Stream.Prelude (foldl', fromList)
+import Streamly.Internal.Data.Stream.Serial (SerialT)
+import Streamly.Internal.Data.Stream.StreamK (IsStream)
+import Streamly.Internal.Data.Time.Units (NanoSecond64(..), toRelTime64)
+
+import qualified Data.List as List
+import qualified Streamly.Internal.Data.Array as Array
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.Stream.IsStream.Lift as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Eliminate as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Nesting as Stream
+import qualified Streamly.Internal.Data.Stream.IsStream.Transform as Stream
+import qualified Streamly.Internal.Data.Stream.StreamK as StreamK
+
+import Prelude hiding (filter, zipWith, concatMap, concat)
+
+------------------------------------------------------------------------------
+-- Sampling
+------------------------------------------------------------------------------
+
+-- | @sampleFromthen offset stride@ samples the element at @offset@ index and
+-- then every element at strides of @stride@.
+--
+-- @
+-- >>> Stream.toList $ Stream.sampleFromThen 3 2 $ S.enumerateFromTo 0 10
+-- > [1,4,7]
```

Comment:
Shouldn't the expected result be, `[2, 4, 6, 8, 10]`?

Reply:
You are right. I just wrote the examples but did not compile/test these. Will fix.